src/pm-conversations/pmConversationsSelectors.js

```diff
@@ -1,46 +1,62 @@
-import { getOwnUser } from '../users/userSelectors';
+import { getRecentPrivateConversations } from '../directSelectors';
+import { getOwnUser, getAllUsersById } from '../users/userSelectors';
 import { getUnreadByPms, getUnreadByHuddles } from '../unread/unreadSelectors';
 import { normalizeRecipientsSansMe, pmUnreadsKeyFromMessage } from '../utils/recipient';
+import { ZulipVersion } from '../utils/zulipVersion';
 
-export const getRecentConversations: Selector<PmConversationData[]> = createSelector(
-  getOwnUser,
-  getPrivateMessages,
-  getUnreadByPms,
-  getUnreadByHuddles,
-  (
-    ownUser: User,
-    messages: Message[],
-    unreadPms: { [number]: number },
-    unreadHuddles: { [string]: number },
-  ): PmConversationData[] => {
-    const recipients = messages.map(msg => ({
-      ids: pmUnreadsKeyFromMessage(msg, ownUser.user_id),
-      emails: normalizeRecipientsSansMe(msg.display_recipient, ownUser.email),
-      msgId: msg.id,
-    }));
+/**
+ * Given a list of PmConversationPartial or PmConversationData, trim it down to
+ * contain only the most recent message from any conversation, and return them
+ * sorted by recency.
+ */
+const collateByRecipient = <T: { recipients: string, msgId: number }>(
+  items: $ReadOnlyArray<T>,
+): T[] => {
+  const latestByRecipient = new Map();
+  items.forEach(item => {
+    const prev = latestByRecipient.get(item.recipients);
+    if (!prev || item.msgId > prev.msgId) {
+      latestByRecipient.set(item.recipients, item);
+    }
+  });
 
-    const latestByRecipient = new Map();
-    recipients.forEach(recipient => {
-      const prev = latestByRecipient.get(recipient.emails);
-      if (!prev || recipient.msgId > prev.msgId) {
-        latestByRecipient.set(recipient.emails, {
-          ids: recipient.ids,
-          recipients: recipient.emails,
-          msgId: recipient.msgId,
-        });
-      }
-    });
+  const sortedByMostRecent = Array.from(latestByRecipient.values()).sort(
+    (a, b) => +b.msgId - +a.msgId,
+  );
+
+  return sortedByMostRecent;
+};
 
-    const sortedByMostRecent = Array.from(latestByRecipient.values()).sort(
-      (a, b) => +b.msgId - +a.msgId,
-    );
+type PmConversationPartial = $Diff<PmConversationData, {| unread: mixed |}>;
+type UnreadAttacher = ($ReadOnlyArray<PmConversationPartial>) => PmConversationData[];
 
-    return sortedByMostRecent.map(recipient => ({
+/**
+ * Auxiliary function: fragment of the selector(s) for PmConversationData.
+ * Transforms PmConversationPartial[] to PmConversationData[].
+ *
+ * Note that, since this will only ever be called by other selectors, the inner
+ * function doesn't need to do any memoization of its own.
+ */
+const getAttachUnread: Selector<UnreadAttacher> = createSelector(
+  getUnreadByPms,
+  getUnreadByHuddles,
+  (unreadPms: { [number]: number }, unreadHuddles: { [string]: number }) => (
+    partials: $ReadOnlyArray<PmConversationPartial>,
+  ) =>
```

Comment:
There's nothing to memoize here, because the body is just a lambda.

The logic inside that lambda -- certainly the expression after `unread:`, complete with comments -- is indeed helpful to factor out into a helper function. That helper can take `unreadPms` and `unreadHuddles` among its arguments.